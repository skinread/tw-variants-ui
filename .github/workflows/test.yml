# This workflow will do a clean installation of node dependencies, cache/restore them, build the source code and run tests across different versions of node
# For more information see: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-nodejs

name: CI Tests

on:
  pull_request:
    branches: ["main","develop"]

jobs:
  test:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        node-version: [20.x]
    steps:
    - name: Checkout
      uses: actions/checkout@v4
    - name: Setup Node ${{ matrix.node-version }}
      uses: actions/setup-node@v4
      with:
        node-version: ${{ matrix.node-version }}
        cache: 'npm'
        cache-dependency-path: "**/package-lock.json"
    - name: Cache or restore dependencies
      uses: actions/cache@v4
      with:
        path: ~/.npm
        key: ${{ runner.os }}-npm-${{ hashFiles('**/package-lock.json') }}
        restore-keys: ${{ runner.os }}-npm-
    - name: Install dependencies
      run: npm ci

    - name: Check formatting and linting
      run: npm run prettier && npm run lint

    - name: Check package build
      run: npm run build

    - name: Get Playwright version
      run: echo "PLAYWRIGHT_VERSION=$(jq -r '.devDependencies["@playwright/test"]' package.json)" >> $GITHUB_ENV
    - name: Cache or restore Playwright binaries
      uses: actions/cache@v4
      with:
        path: |
          ~/.cache/ms-playwright
        key: ${{ runner.os }}-playwright-${{ env.PLAYWRIGHT_VERSION }}
        restore-keys: ${{ runner.os }}-playwright-
    - name: Install Playwright
      if: steps.playwright-cache.outputs.cache-hit != 'true'
      run: npm run pretest
    - name: Storybook test build
      run: npm run storybook:build -- --test
    - name: Storybook test suite
      run: npm run test:ci
