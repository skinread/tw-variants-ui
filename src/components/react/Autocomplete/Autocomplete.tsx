import { forwardRef, useRef, useState } from 'react';
import { Combobox } from '@headlessui/react';
import { autocompleteVariants, validityFromProps } from '@common';
import { type InputFieldProps } from '@components/react';
import { FieldFeedback } from '../InputField/FieldFeedback';

export interface AutocompleteProps
  extends Omit<InputFieldProps, 'type' | 'localise' | 'dir' | 'value'> {
  /**
   * Options data
   */
  options: {
    label: string;
    value: string;
  }[];
  /**
   * Custom callback with current value instead of event
   */
  onChangeValue?: (value: string) => void;
  /**
   * Only use value for a controlled input scenario
   */
  value?: InputFieldProps['value'] | null;
}
// we extract the options from the interface because Storybook Autodocs doesn't expand into nested types
type ArrayType<T> = T extends (infer U)[] ? U : never;
export type Options = AutocompleteProps['options'];
export type Option = ArrayType<Options>;

/**
 * Simple implementation of an autocompletion combobox field. Start typing to filter options. React `ref` enabled.
 */
export const Autocomplete = forwardRef<HTMLInputElement, AutocompleteProps>(
  function VuiAutocomplete(props, ref) {
    const {
      defaultValue,
      descriptionContent,
      disabled,
      feedbackContent,
      hasError,
      id: _id,
      infoContent,
      label,
      onChangeValue,
      options,
      value,
      ...elemAttrs
    } = props;
    const {
      wrapper: wrapperStyle,
      label: labelStyle,
      labelText: labelTextStyle,
      input: inputStyle,
      menu: menuStyle,
      item: itemStyle,
      descriptionWrapper: descriptionWrapperStyle,
    } = autocompleteVariants({ type: 'default', validity: validityFromProps({ hasError }) });

    const id = useRef(_id ?? `autocomplete-${props.name}`);
    const idFeedback = `${id.current}-feedback`;
    const idDescription = `${id.current}-description`;
    const [query, setQuery] = useState('');

    const getLabelFromValue = (value: string) => {
      const option = options.find((o) => o.value === value);
      return option?.label ?? '';
    };

    const filteredOptions = () => {
      if (!query.length) return options;

      const filtered = options.filter((item) => {
        // if query is 1 - 3 characters match only the start of the words or value
        if (query.length < 3)
          return (
            item.value.toLowerCase().startsWith(query.toLowerCase()) ||
            !!item.label
              .toLowerCase()
              .split(' ')
              .find((word) => word.startsWith(query.toLowerCase()))
          );
        // else match any characters
        return item.label.toLowerCase().includes(query.toLowerCase());
      });

      return filtered.length ? filtered : options;
    };

    const Description = () => {
      if (!descriptionContent) return null;
      return (
        <div className={descriptionWrapperStyle()} id={idDescription}>
          {descriptionContent}
        </div>
      );
    };

    return (
      <Combobox
        onChange={(value: string) => {
          setQuery(getLabelFromValue(value));
          onChangeValue?.(value);
        }}
        defaultValue={defaultValue}
        disabled={disabled}
        value={value}
        nullable
      >
        <div className={wrapperStyle()}>
          <Combobox.Label className={labelStyle()} htmlFor={id.current}>
            <span className={labelTextStyle()}>{label}</span>
            {infoContent && <span>{infoContent}</span>}
          </Combobox.Label>
          <div className="relative">
            {/**
             * // can be wrapped in the button component to auto-open the menu on focus
             * <Combobox.Button className="w-full">
             **/}
            <Combobox.Input
              className={inputStyle()}
              id={id.current}
              aria-labelledby={feedbackContent ? idFeedback : undefined}
              aria-describedby={descriptionContent ? idDescription : undefined}
              displayValue={(value: string) => getLabelFromValue(value)}
              onChange={(e) => {
                setQuery(e.currentTarget.value);
              }}
              {...elemAttrs}
              ref={ref}
            />
            {/* </Combobox.Button> */}
            <Combobox.Options className={menuStyle()}>
              {filteredOptions().map((option) => (
                <Combobox.Option
                  className={({ active }) => itemStyle({ class: active && 'bg-neutral-focus' })}
                  value={option.value}
                  key={option.value}
                >
                  {option.label}
                </Combobox.Option>
              ))}
            </Combobox.Options>
          </div>
          <FieldFeedback
            feedbackContent={feedbackContent}
            idFeedback={idFeedback}
            hasError={hasError}
          />
          <Description />
        </div>
      </Combobox>
    );
  }
);
