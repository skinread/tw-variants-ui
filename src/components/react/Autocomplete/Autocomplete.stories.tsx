import type { Meta, StoryObj } from '@storybook/react';
import { expect, userEvent, within } from '@storybook/test';

import { Autocomplete, type Option } from './Autocomplete';

const fruits = [
  {
    label: 'Apple',
    value: 'a',
  },
  {
    label: 'Banana',
    value: 'b',
  },
  {
    label: 'Cherries',
    value: 'c',
  },
  {
    label: 'Dragon Fruit',
    value: 'e',
  },
  {
    label: 'Grapefruit',
    value: 'g',
  },
  {
    label: 'Mango',
    value: 'm',
  },
  {
    label: 'Orange',
    value: 'o',
  },
  {
    label: 'Kiwi',
    value: 'k',
  },
  {
    label: 'Rasperries',
    value: 'r',
  },
];

const meta = {
  title: 'React/Autocomplete',
  component: Autocomplete,
  tags: ['autodocs'],
} satisfies Meta<typeof Autocomplete>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Fruit: Story = {
  args: {
    label: 'Fruit',
    options: fruits,
    name: 'demo-fruit-field',
    placeholder: 'Start typing',
  },
};

export const Invalid: Story = {
  args: {
    ...Fruit.args,
    hasError: true,
    feedbackContent: 'Fruit is required',
  },
};

/**
 * Open this story directly to observe automated test
 */
export const InteractionTest: Story = {
  args: Fruit.args,
  decorators: [
    (Story) => (
      <form>
        <Story />
      </form>
    ),
  ],
  play: async ({ args, canvasElement, step }) => {
    const canvas = within(canvasElement);
    const input = canvas.getByLabelText<HTMLInputElement>(args.label, { selector: 'input' });

    async function doSelectOption(input: HTMLInputElement, option: Option) {
      const { label, value } = option;
      await userEvent.type(input, value, { delay: 50 });
      await userEvent.keyboard('[ArrowDown][Enter]', { delay: 50 });
      await expect(input).toHaveValue(label);
    }

    let option = args.options[1];
    await step(`Select ${option?.label}`, async () => {
      await doSelectOption(input, option!);
    });

    await step('Clear field', async () => {
      await userEvent.clear(input);
      await expect(input).toHaveValue('');
    });

    option = args.options[0];
    await step(`Select ${option?.label}`, async () => {
      await doSelectOption(input, option!);
    });
  },
};
