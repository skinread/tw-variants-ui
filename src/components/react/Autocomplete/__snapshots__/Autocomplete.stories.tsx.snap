// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React/Autocomplete Fruit smoke-test 1`] = `
<div class="form-control">
  <label class="label items-start px-0"
         for="autocomplete-demo-fruit-field"
         id="headlessui-combobox-label-:r0:"
         data-headlessui-state
  >
    <span class="label-text text-base font-bold">
      Fruit
    </span>
  </label>
  <div class="relative">
    <input class="placeholder:text-subtle w-full input input-primary rounded-[--rounded-field] h-[42px] px-3"
           name="demo-fruit-field"
           placeholder="Start typing"
           id="autocomplete-demo-fruit-field"
           role="combobox"
           type="text"
           aria-expanded="false"
           aria-autocomplete="list"
           data-headlessui-state
           value
           aria-labelledby="headlessui-combobox-label-:r0:"
    >
  </div>
</div>
`;

exports[`React/Autocomplete InteractionTest play-test 1`] = `
<form>
  <div class="form-control">
    <label class="label items-start px-0"
           for="autocomplete-demo-fruit-field"
           id="headlessui-combobox-label-:r6:"
           data-headlessui-state
    >
      <span class="label-text text-base font-bold">
        Fruit
      </span>
    </label>
    <div class="relative">
      <input class="placeholder:text-subtle w-full input input-primary rounded-[--rounded-field] h-[42px] px-3"
             name="demo-fruit-field"
             placeholder="Start typing"
             id="autocomplete-demo-fruit-field"
             role="combobox"
             type="text"
             aria-expanded="false"
             aria-autocomplete="list"
             data-headlessui-state
             value
             aria-labelledby="headlessui-combobox-label-:r6:"
      >
    </div>
  </div>
</form>
`;

exports[`React/Autocomplete Invalid smoke-test 1`] = `
<div class="form-control">
  <label class="label items-start px-0"
         for="autocomplete-demo-fruit-field"
         id="headlessui-combobox-label-:r3:"
         data-headlessui-state
  >
    <span class="label-text text-base font-bold">
      Fruit
    </span>
  </label>
  <div class="relative">
    <input class="placeholder:text-subtle w-full input input-primary rounded-[--rounded-field] border-2 h-[42px] px-3 input-error"
           name="demo-fruit-field"
           placeholder="Start typing"
           id="autocomplete-demo-fruit-field"
           role="combobox"
           type="text"
           aria-expanded="false"
           aria-autocomplete="list"
           data-headlessui-state
           value
           aria-labelledby="headlessui-combobox-label-:r3:"
    >
  </div>
  <label class="leading-5 mt-2 [&amp;>label]:mt-0 text-error fill-error flex"
         id="autocomplete-demo-fruit-field-feedback"
  >
    <div class="-mt-0.5">
      <svg xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 24 24"
           fill="currentColor"
           class="inline-block  w-6 h-6 mr-1.5 mb-0.5 icon-exclamation"
           aria-hidden="true"
      >
        <path d="m22.242,5.272l-3.515-3.515c-1.133-1.133-2.64-1.757-4.242-1.757h-4.971c-1.603,0-3.109.624-4.242,1.757l-3.515,3.514c-1.134,1.133-1.758,2.64-1.758,4.243v4.971c0,1.603.624,3.11,1.758,4.243l3.515,3.515c1.133,1.133,2.64,1.757,4.242,1.757h4.971c1.603,0,3.109-.624,4.242-1.757l3.515-3.514c1.134-1.133,1.758-2.64,1.758-4.243v-4.971c0-1.603-.624-3.11-1.758-4.243Zm-.242,9.213c0,1.069-.416,2.073-1.172,2.829l-3.515,3.515c-.756.755-1.76,1.171-2.828,1.171h-4.971c-1.068,0-2.072-.416-2.828-1.171l-3.515-3.515c-.756-.755-1.172-1.759-1.172-2.828v-4.971c0-1.069.416-2.073,1.172-2.829l3.515-3.515c.756-.755,1.76-1.171,2.828-1.171h4.971c1.068,0,2.072.416,2.828,1.171l3.515,3.515c.756.755,1.172,1.759,1.172,2.828v4.971Zm-11-.485V6c0-.552.447-1,1-1s1,.448,1,1v8c0,.552-.447,1-1,1s-1-.448-1-1Zm2.5,3.5c0,.828-.672,1.5-1.5,1.5s-1.5-.672-1.5-1.5.672-1.5,1.5-1.5,1.5.672,1.5,1.5Z"
              stroke="none"
        >
        </path>
      </svg>
    </div>
    Fruit is required
  </label>
</div>
`;
